### DISCUSSION TITLE
Interaction restriction

### DISCUSSION THREAD

**Discussion by luiswirth** [emilk/egui#346] [Q&A] [answered]
I'm currently developing my own drawing application, which uses egui.
I have a window in which I render the pen drawings of the user. This is rendered directly with wgpu (no epaint) and egui just shows an image, which is the render target.

Drawing is done using the mouse. When a mouse button is pressed down it is supposed to draw. Unfortunately currently egui moves the window when the mouse "drags" it. How can I prevent this? I don't want the window to be fixed in place, but only moveable by the borders.

This problem is probably part of a broader one: Being able to customize how the user can interact with egui.

**Comment by emilk** [comment, marked as answer]
There is a painting demo at https://emilk.github.io/egui/#demo with the source code at https://github.com/emilk/egui/blob/master/egui_demo_lib/src/apps/demo/painting.rs

The critical part you are missing is `Sense::drag()` - this tells egui that you will be catching drags, and that egui should not use them

**Reply by luiswirth** [reply]
So, I would also need to allocate something. I believe it should be `Ui::allocate_ui`, since I want to have an image (render target) inside of it. But this doesn't let me specify a `Sense`.

**Reply by emilk** [reply]
If you use `allocate_painter` you get an `egui::Painter` which allows you to do low-level drawing, which sounds like what you want.

There is also `allocate_response`: https://docs.rs/egui/0.11.0/egui/struct.Ui.html#method.allocate_response which gives you a `Response` object that allows you to check interactions (`if response.dragged() { … }`). You can paint with `ui.painter()`.

**Reply by luiswirth** [reply]
I don't want to use the `ui.painter()`. I want to use my own rendering pipeline (including my own tessellation pipeline). This is why I'm just showing an image, which is the texture render target of my own pipeline.

**Reply by luiswirth** [reply]
This is because I am developing a drawing application, which should also be able to render stroked paths and a lot more, which `epaint` just doesn't provide. Which of course is fine, that's not the purpose of `epaint`. But I need more control of the window I'm rendering in. So I'm using my own pipeline.

**Reply by luiswirth** [reply]
The step of hooking up the custom render pipeline with egui is render-backend specific. I've already implemented this as [`egui_wgpu::Rendertarget`](https://docs.rs/egui-wgpu/0.0.0-alpha.3/egui_wgpu/struct.RenderTarget.html).

The interaction system however is something that would have to be part of egui I believe. It sorta is platform (e.g. winit) specify, but egui needs to provide some sense of a window related event, which can be forwarded/delegated to my own system.

**Reply by luiswirth** [reply]
I think this is something very valuable for egui. Since it advertises itself as a gui framework for game engines.

Think of Unity, where the scene/game view is just a window. This setup also requires the gui to be hooked up to a custom rendering pipeline and a interaction system, which forwards the events specific to this window to some kind of scene handler.

**Reply by emilk** [reply]
Rendering custom content within an egui Ui would indeed be great, and the only way to do so currently is via a texture/image.

I just pushed this: https://github.com/emilk/egui/commit/d862ff66acdfc5c9f9df4007301c07acdaca91d1

now you can make an `egui::Image` "swallow" input with `image.sense(Sense::click_and_drag())`

**Reply by luiswirth** [reply]
Oh great, thanks! I will try it out, when I have time.

**Reply by luiswirth** [reply]
Nice, it works for me :).

**Reply by luiswirth** [reply]
For now this question is resolved. But I think it would be nice to really forward the events and transform the positions etc to be relative to the ui element.
